### PR TITLE
[IRGen] Fix witness table miscompilation.

### PR DESCRIPTION
**Pull request by al45tair**
If a foreign type conforms to a protocol, and that conformance is not resilient, we were emitting an instantantiation function but it would never be called because we weren't emitting the `GenericWitnessTable` that would have contained a reference to it.

This was happening because of a missing `isSynthesizedNonUnique()` call in `isDependentConformance()`.

rdar://97290618

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
Hmmm. That exploded unexpectedly. I'll keep digging then.

File: lib/IRGen/GenProto.cpp [hunk 1]
```diff
@@ -961,11 +967,13 @@ static bool isDependentConformance(
     if (assocConformance.isInvalid())
       return false;
 
+    auto rootConformance = assocConformance.getConcrete()->getRootConformance();
```
Comment (aschwaighofer) [inline lib/IRGen/GenProto.cpp:970]: I think you need to guard this by `assocConformance.isAbstract()`

**Comment by aschwaighofer** [comment]
Is it possible to add a test case?

File: lib/IRGen/GenProto.cpp (hunk 1 above)
Comment (aschwaighofer) [reply lib/IRGen/GenProto.cpp:970]: ... `! assocConformance.isAbstract()`

Comment (al45tair) [reply lib/IRGen/GenProto.cpp:973]: Yes, I figured out as much after doing a local build.

**Comment by al45tair** [comment]
> Is it possible to add a test case?

Should be, yes.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Review by rjmccall — CHANGES_REQUESTED** [review]
LGTM except that it's missing a test case.

**Comment by al45tair** [comment]
Having looked at this a bit further, I don't think it's feasible to write a test case that will work on macOS (this only affects synthesized conformances, which means the protocol in question needs to be in the Swift standard library, but on macOS that's dynamically linked and always resilient; building a test case there would require a fairly heroic effort), but it's probably possible on Linux. Trying to get a Linux build up so I can test it.

**Comment by al45tair** [comment]
> Having looked at this a bit further, I don't think it's feasible to write a test case that will work on macOS

It turns out that it _was_ feasible. Just somewhat tricky to get right (the compiler doesn't give very helpful error messages when bits of the `Swift` module are missing; it often just crashes somewhere instead).

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Review by aschwaighofer — APPROVED** [review]
lgtm

**Comment by al45tair** [comment]
Failed on Windows and Linux, probably because of the section name. Looking at that now.

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Comment by al45tair** [comment]
@swift-ci Please smoke test

**Review by rjmccall — APPROVED** [review]
LGTM

**Pull request merged**